--- a/pkg/tcpip/adapters/gonet/gonet.go
+++ b/pkg/tcpip/adapters/gonet/gonet.go
@@ -459,6 +459,11 @@ func (c *TCPConn) RemoteAddr() net.Addr {
 	return fullToTCPAddr(a)
 }
 
+// Endpoint returns the underlying tcpip.Endpoint.
+func (c *TCPConn) Endpoint() tcpip.Endpoint {
+	return c.ep
+}
+
 func (c *TCPConn) newOpError(op string, err error) *net.OpError {
 	return &net.OpError{
 		Op:     op,
@@ -711,3 +716,8 @@ func (c *UDPConn) LocalAddr() net.Addr {
 	}
 	return fullToUDPAddr(a)
 }
+
+// Endpoint returns the underlying tcpip.Endpoint.
+func (c *UDPConn) Endpoint() tcpip.Endpoint {
+	return c.ep
+}
